core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTypeUtil.java

```diff
@@ -74,6 +81,12 @@
                     .put(Object.class, Function.identity())
                     .build();
 
+    /**
+     * "Cache" of converters from string to the Class in the key.
+     */
+    private static final Map<Class<?>, Function<String, ?>> convertExternalTypes =
+            new MapMaker().weakKeys().weakValues().makeMap();
```

Comment:
I guess the value doesn't need to be weak. The value will be automatically deferenced when the key is collected.

Reply:
But is the key really collected, or just the value of the WeakRef? That question is something I was unsure enough of to just go with weakValues().

Reply:
The value of the `WeakReference` will be collected. If the map implementation finds out that the `WeakReference` is empty, the map entry will be cleaned up. I think that's how all weak key maps work in Java. `weakValues()` basically does the same thing; the only difference is that it does that on an entry value, i.e. if an entry value's `WeakReference` is empty, it will clean up the entry.

Reply:
So, in this case, we do not want the entry to be cleaned up when the value is GC'd, because it's very likely that the value (`Function`) is only referred via this map (`convertExternalTypes`).

Reply:
Very true, didn't think about that case. Will remove `weakValues()`

Reply:
Fixed